src/v/storage/disk_log_impl.cc

```diff
@@ -1139,23 +1139,36 @@ ss::future<> disk_log_impl::do_truncate(truncate_config cfg) {
     }
 
     co_await last->flush();
+    /**
+     * We look for the offset preceding the the requested truncation offset.
+     *
+     * This guarantee that the offset to file position translating consumer will
+     * see at least one batch that precedes the truncation point. This will
+     * allow establishing correct dirty offset after truncation.
+     */
 
-    auto pidx = last->index().find_nearest(cfg.base_offset);
-    model::offset start = last->index().base_offset();
+    // if no offset is found in an index we will start from the segment base
+    // offset
+    model::offset start = last->offsets().base_offset;
+
+    auto pidx = last->index().find_nearest(
+      std::max(start, model::prev_offset(cfg.base_offset)));
     size_t initial_size = 0;
     if (pidx) {
         start = pidx->offset;
         initial_size = pidx->filepos;
     }
+
     auto initial_generation_id = last->get_generation_id();
 
     // an unchecked reader is created which does not enforce the logical
     // starting offset. this is needed because we really do want to read
     // all the data in the segment to find the correct physical offset.
     auto reader = co_await make_unchecked_reader(
-      log_reader_config(start, cfg.base_offset, cfg.prio));
+      log_reader_config(start, model::offset::max(), cfg.prio));
```

Comment:
what scenario requires reading past base offset?

Reply:
we need to read past the base offset to be able to read next batch after a gap